sm2/src/pke.rs

```diff
@@ -0,0 +1,178 @@
+use alloc::vec;
+
+use elliptic_curve::{
+    bigint::{Encoding, Uint, U256},
+    pkcs8::der::{
+        asn1::UintRef, Decode, DecodeValue, Encode, Length, Reader, Sequence, Tag, Writer,
+    },
+};
+
+use elliptic_curve::{
+    pkcs8::der::{asn1::OctetStringRef, EncodeValue},
+    sec1::ToEncodedPoint,
+    Result,
+};
+use sm3::digest::DynDigest;
+
+#[cfg(feature = "arithmetic")]
+mod decrypting;
+#[cfg(feature = "arithmetic")]
+mod encrypting;
+
+#[cfg(feature = "arithmetic")]
+pub use self::{decrypting::DecryptingKey, encrypting::EncryptingKey};
+
+/// Modes for the cipher encoding/decoding.
+#[derive(Clone, Copy, Debug)]
+pub enum Mode {
+    /// old mode
+    C1C2C3,
+    /// new mode
+    C1C3C2,
+}
+/// Represents a cipher structure containing encryption-related data (asn.1 format).
+///
+/// The `Cipher` structure includes the coordinates of the elliptic curve point (`x`, `y`),
+/// the digest of the message, and the encrypted cipher text.
+pub struct Cipher<'a> {
+    x: U256,
+    y: U256,
+    digest: &'a [u8],
+    cipher: &'a [u8],
+}
+
+impl<'a> Sequence<'a> for Cipher<'a> {}
+
+impl<'a> EncodeValue for Cipher<'a> {
+    fn value_len(&self) -> elliptic_curve::pkcs8::der::Result<Length> {
+        UintRef::new(&self.x.to_be_bytes())?.encoded_len()?
+            + UintRef::new(&self.y.to_be_bytes())?.encoded_len()?
+            + OctetStringRef::new(self.digest)?.encoded_len()?
+            + OctetStringRef::new(self.cipher)?.encoded_len()?
+    }
+
+    fn encode_value(&self, writer: &mut impl Writer) -> elliptic_curve::pkcs8::der::Result<()> {
+        UintRef::new(&self.x.to_be_bytes())?.encode(writer)?;
+        UintRef::new(&self.y.to_be_bytes())?.encode(writer)?;
+        OctetStringRef::new(self.digest)?.encode(writer)?;
+        OctetStringRef::new(self.cipher)?.encode(writer)?;
+        Ok(())
+    }
+}
+
+impl<'a> DecodeValue<'a> for Cipher<'a> {
+    type Error = elliptic_curve::pkcs8::der::Error;
+
+    fn decode_value<R: Reader<'a>>(
+        decoder: &mut R,
+        header: elliptic_curve::pkcs8::der::Header,
+    ) -> core::result::Result<Self, Self::Error> {
+        decoder.read_nested(header.length, |nr| {
+            let x = UintRef::decode(nr)?.as_bytes();
+            let y = UintRef::decode(nr)?.as_bytes();
+            let digest = OctetStringRef::decode(nr)?.into();
+            let cipher = OctetStringRef::decode(nr)?.into();
+            Ok(Cipher {
+                x: Uint::from_be_bytes(zero_pad_byte_slice(x)?),
+                y: Uint::from_be_bytes(zero_pad_byte_slice(y)?),
+                digest,
+                cipher,
+            })
+        })
+    }
+}
+
+/// Performs key derivation using a hash function and elliptic curve point.
+fn kdf(hasher: &mut dyn DynDigest, kpb: AffinePoint, c2: &mut [u8]) -> Result<()> {
+    let klen = c2.len();
+    let mut ct: i32 = 0x00000001;
+    let mut offset = 0;
+    let digest_size = hasher.output_size();
+    let mut ha = vec![0u8; digest_size];
```

Comment:
This won't work on heapless `no_std` targets, nor is the `alloc` feature being enabled by the `pke` feature.

It seems like you could get rid of these dynamic allocations if you switched from `DynDigest` to a generic `D` parameter like you have for `decrypt_digest` and bound on the [`OutputSizeUser`](https://docs.rs/digest/latest/digest/core_api/trait.OutputSizeUser.html) trait, which will give you access to a compile-time output size which can be used to allocate an appropriately sized array.

Reply:
> I once saw an [issue](https://github.com/RustCrypto/traits/issues/1574) related to using `OutputSizeUser::USIZE` to create a new array.

I tried to solve this problem with the [`hybird_array`](https://github.com/RustCrypto/hybrid-array/), but I needed a little time to adjust the other places where the `alloc` feature was used